File: frontend/src/app/shared/services/maintenance.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface MaintenanceRecord {
  id?: string;
  vehicleId: string;
  vehicleName?: string;
  type: string;
  description?: string;
  cost?: number;
  odometerKm?: number;
  performedAt: string;
  nextDueDate?: string;
  nextDueKm?: number;
  performedBy?: string;
  notes?: string;
}

@Injectable({ providedIn: 'root' })
export class MaintenanceService {
  private http = inject(HttpClient);
  private api = `${environment.apiUrl}/maintenance`;

  getAll(): Observable<MaintenanceRecord[]> { return this.http.get<MaintenanceRecord[]>(this.api); }
  getByVehicle(vehicleId: string): Observable<MaintenanceRecord[]> { return this.http.get<MaintenanceRecord[]>(`${this.api}/vehicle/${vehicleId}`); }
  create(record: MaintenanceRecord): Observable<MaintenanceRecord> { return this.http.post<MaintenanceRecord>(this.api, record); }
  update(id: string, record: MaintenanceRecord): Observable<MaintenanceRecord> { return this.http.put<MaintenanceRecord>(`${this.api}/${id}`, record); }
  delete(id: string): Observable<void> { return this.http.delete<void>(`${this.api}/${id}`); }
}

File: backend/src/main/java/eu/fuelfleet/maintenance/service/MaintenanceService.java
package eu.fuelfleet.maintenance.service;

import eu.fuelfleet.auth.entity.User;
import eu.fuelfleet.auth.repository.UserRepository;
import eu.fuelfleet.maintenance.dto.MaintenanceRequest;
import eu.fuelfleet.maintenance.dto.MaintenanceResponse;
import eu.fuelfleet.maintenance.entity.MaintenanceRecord;
import eu.fuelfleet.maintenance.repository.MaintenanceRepository;
import eu.fuelfleet.vehicle.entity.Vehicle;
import eu.fuelfleet.vehicle.repository.VehicleRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class MaintenanceService {
    private final MaintenanceRepository maintenanceRepository;
    private final VehicleRepository vehicleRepository;
    private final UserRepository userRepository;

    public List<MaintenanceResponse> getAll() {
        return maintenanceRepository.findByCompanyIdOrderByPerformedAtDesc(getCompanyId()).stream()
                .map(MaintenanceResponse::from).toList();
    }

    public List<MaintenanceResponse> getByVehicle(UUID vehicleId) {
        return maintenanceRepository.findByVehicleIdOrderByPerformedAtDesc(vehicleId).stream()
                .map(MaintenanceResponse::from).toList();
    }

    @Transactional
    public MaintenanceResponse create(MaintenanceRequest req) {
        User user = getCurrentUser();
        Vehicle vehicle = vehicleRepository.findById(req.getVehicleId())
                .orElseThrow(() -> new RuntimeException("Vehicle not found"));
        MaintenanceRecord record = MaintenanceRecord.builder()
                .vehicle(vehicle).company(user.getCompany())
                .type(req.getType()).description(req.getDescription())
                .cost(req.getCost()).odometerKm(req.getOdometerKm())
                .performedAt(req.getPerformedAt()).nextDueDate(req.getNextDueDate())
                .nextDueKm(req.getNextDueKm()).performedBy(req.getPerformedBy())
                .notes(req.getNotes())
                .build();
        return MaintenanceResponse.from(maintenanceRepository.save(record));
    }

    @Transactional
    public MaintenanceResponse update(UUID id, MaintenanceRequest req) {
        MaintenanceRecord r = maintenanceRepository.findById(id)
                .filter(m -> m.getCompany().getId().equals(getCompanyId()))
                .orElseThrow(() -> new RuntimeException("Record not found"));
        if (req.getVehicleId() != null) {
            Vehicle v = vehicleRepository.findById(req.getVehicleId()).orElseThrow();
            r.setVehicle(v);
        }
        r.setType(req.getType()); r.setDescription(req.getDescription());
        r.setCost(req.getCost()); r.setOdometerKm(req.getOdometerKm());
        r.setPerformedAt(req.getPerformedAt()); r.setNextDueDate(req.getNextDueDate());
        r.setNextDueKm(req.getNextDueKm()); r.setPerformedBy(req.getPerformedBy());
        r.setNotes(req.getNotes());
        return MaintenanceResponse.from(maintenanceRepository.save(r));
    }

    @Transactional
    public void delete(UUID id) {
        maintenanceRepository.findById(id)
                .filter(m -> m.getCompany().getId().equals(getCompanyId()))
                .ifPresent(maintenanceRepository::delete);
    }

    private UUID getCompanyId() { return getCurrentUser().getCompany().getId(); }
    private User getCurrentUser() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return userRepository.findByEmail(email).orElseThrow();
    }
}
